packages/langium/src/utils/grammar-util.ts

```diff
@@ -281,59 +269,286 @@ function findNameAssignmentInternal(type: ast.AbstractType, cache: Map<ast.Abstr
+        return negateTokenToRegex(element);
+    } else if (ast.isUntilToken(element)) {
+        return untilTokenToRegex(element);
+    } else if (ast.isRegexToken(element)) {
+        const lastSlash = element.regex.lastIndexOf('/');
+        const source = element.regex.substring(1, lastSlash);
+        const regexFlags = element.regex.substring(lastSlash + 1);
+        if (flags) {
+            flags.i = regexFlags.includes('i');
+            flags.s = regexFlags.includes('s');
+            flags.u = regexFlags.includes('u');
+        }
+        return withCardinality(source, {
+            cardinality: element.cardinality,
+            lookahead: element.lookahead,
+            wrap: false
+        });
+    } else if (ast.isWildcard(element)) {
+        return withCardinality(WILDCARD, {
+            cardinality: element.cardinality,
+            lookahead: element.lookahead
+        });
+    } else {
+        throw new Error(`Invalid terminal element: ${element?.$type}`);
+    }
+}
+
+function terminalAlternativesToRegex(alternatives: ast.TerminalAlternatives): string {
+    return withCardinality(alternatives.elements.map(e => abstractElementToRegex(e)).join('|'), {
+        cardinality: alternatives.cardinality,
+        lookahead: alternatives.lookahead
+    });
+}
+
+function terminalGroupToRegex(group: ast.TerminalGroup): string {
+    return withCardinality(group.elements.map(e => abstractElementToRegex(e)).join(''), {
+        cardinality: group.cardinality,
+        lookahead: group.lookahead
+    });
+}
+
+function untilTokenToRegex(until: ast.UntilToken): string {
+    return withCardinality(`${WILDCARD}*?${abstractElementToRegex(until.terminal)}`, {
+        cardinality: until.cardinality,
+        lookahead: until.lookahead
+    });
+}
+
+function negateTokenToRegex(negate: ast.NegatedToken): string {
+    return withCardinality(`(?!${abstractElementToRegex(negate.terminal)})${WILDCARD}*?`, {
+        cardinality: negate.cardinality,
+        lookahead: negate.lookahead
+    });
+}
+
+function characterRangeToRegex(range: ast.CharacterRange): string {
+    if (range.right) {
+        return withCardinality(`[${keywordToRegex(range.left)}-${keywordToRegex(range.right)}]`, {
+            cardinality: range.cardinality,
+            lookahead: range.lookahead,
+            wrap: false
+        });
+    }
+    return withCardinality(keywordToRegex(range.left), {
+        cardinality: range.cardinality,
+        lookahead: range.lookahead,
+        wrap: false
+    });
+}
+
+function keywordToRegex(keyword: ast.Keyword): string {
+    return escapeRegExp(keyword.value);
+}
+
+function withCardinality(regex: string, options: {
+    cardinality?: string
+    wrap?: boolean
+    lookahead?: string
+}): string {
+    if (options.wrap !== false || options.lookahead) {
+        regex = `(${options.lookahead ?? ''}${regex})`;
+    }
+    if (options.cardinality) {
+        return `${regex}${options.cardinality}`;
+    }
+    return regex;
```

Comment:
All these functions are exposed as top-level exports. I'm thinking about wrapping them in a namespace so it's clear what they are about. WDYT?

Reply:
Let's do this in a follow-up PR as discussed.